### ISSUE TITLE
Use `dataobj` instead of `get_data()` for atlas querying

### ISSUE THREAD

**Issue by rmarkello** [miykael/atlasreader#38]
As of #35, any atlases needed for the analysis are loaded into memory (by calling `atlas.get_data()`) once, towards the beginning of a workflow (e.g., `get_statmap_info()` or `create_output()`). By ensuring this is only done once we saves a lot of time, since loading the data is quite time-consuming (we were previously loading the data on _every single loop_, so, hurray for that fix).

However, this entire process could be pretty dramatically sped up by never loading the atlases into memory at all (that is, never calling `atlas.get_data()`). Instead, we could replace that by simply accessing the needed data via the `atlas.dataobj` array proxy. This might take a bit of finessing since the `dataobj` object only supports integer based indexing (i.e., `atlas.dataobj[0, 0, 0]`, not `atlas.dataobj[((0, 0, 0), (1, 1, 1), (2, 2, 2))]`). Still, I think it might be a worthwhile endeavor&mdash;or, at the very least, something to look into a bit more.

**Comment by miykael** [comment]
Could we also just use a `nibabel` `memmap` approach? Or is that kind of what you mean?

Also, if you need to feed in every voxel individually, don't you think that this will take forever (in a normal loop).

Perhaps we could also look into `numba`, `cython` or `dask` as a way to speeding this up?

**Comment by miykael** [comment]
Hmm.. I quick profiling of the code with `%prun` also showed that the decompression of the atlases takes the longest.
So, I checked: Compressed, the files are a total of 15MB. Uncompressed 1.5GB!!! Mostly due to `atlas_juelich.nii` of 873.3MB and `atlas_harvard_oxford.nii` 469.4MB. Don't know what to do out of this.

**Comment by rmarkello** [comment]
Yes, the `memmap` approach!! Objects loaded with `nibabel` generally have a `dataobj` attribute, which is an `ArrayProxy` that allows for slicing of the underlying array without having to load the whole thing into memory (which is a very costly procedure).

The line you referenced is simply loading the objects into memory -- you'll notice that most images loaded with `nibabel` also have a boolean `in_memory` attribute. It starts off as `False`, but as soon as `img.get_data()` is called once, it becomes `True` because `nibabel` caches the data; subsequent calls to `img.get_data()` are immediate because they're simply returning the cached array. You can use `img.uncache()` to get rid of that; the next time you call `img.get_data()` will thus have to re-read the data and be much slower.

I think I'll be able to coerce things into using the `dataobj` to our advantage so we don't ever have to call `get_data()` -- give me a couple of days to try and turn something around to that effect. I know I gave us a deadline of September 1st for the first release, so I'll see what I can do!

**Comment by miykael** [comment]
Sorry, still reflecting: Do the atlases still work even if we crop them a bit? Many of them have a lot of "zeros" around the ROIs. So perhaps we could use `from nilearn.image import crop_img` to reduce image size?

I did some benchmarking on loading the data from the Juelich atlas (compressed and uncompressed, as well as original and cropped) - **Focus on the loading time on the right**:
```
# Compressed files (15MB)
img: normal;  compressed: True; data_type = 'char' - Loads in 3.83 s
img: normal;  compressed: True; data_type = 'short' - Loads in 7.72 s
img: normal;  compressed: True; data_type = 'int' - Loads in 15.8 s
img: normal;  compressed: True; data_type = 'float' - Loads in 15.9 s

img: cropped; compressed: True; data_type = 'char' - Loads in 2.2 s
img: cropped; compressed: True; data_type = 'short' - Loads in 4.04 s

# Uncompressed files (but 1.5GB)
img: normal;  compressed: False; data_type = 'char' - Loads in 50.1 ms
img: normal;  compressed: False; data_type = 'short' - Loads in 2.54 ms
img: normal;  compressed: False; data_type = 'int' - Loads in 2.99 ms
img: normal;  compressed: False; data_type = 'float' - Loads in 2.55 ms

img: cropped; compressed: False; data_type = 'char' - Loads in 2.69 ms
img: cropped; compressed: False; data_type = 'short' - Loads in 2.82 ms
```

All this to say: Yes, we need a more direct access of the atlas data!

**Comment by rmarkello** [comment]
Nice! That ~halving in loading time for compressed files is 💯. If the `dataobj` accessing fails then I think cropping should be our go-to 👍.

**Comment by rmarkello** [comment]
Sorry I never responded to this&mdash;I don't know how I missed it! Anyway, I think this is resolved with #67. Moreover, local tests of accessing the `dataobj` directly for `.nii.gz`files seems to offer only limited improvements in speed, so I'm closing this for now!
